### PR TITLE
CRM-49 Create addresses table

### PR DESCRIPTION
**Pull request by aliwocha**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Commits are suqashed
- [x] PR and commit messages follow the style guidelines

### CODE REVIEW

File: backend/src/main/java/com/crm/model/db/AddressEntity.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    @NotBlank
+    @Size(max = 30)
+    private String city;
+
+    @NotBlank
+    @Size(max = 30)
+    private String postalCode;
+
+    @NotBlank
+    @Size(max = 30)
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/model/db/AddressEntity.java:47]: The longest street name in poland has 56 characters. Maybe consider this in size of street name.
https://tvn24.pl/polska/tna-nazwy-ulic-z-oszczednosci-ra292263-3515213

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java [hunk 2]
```diff
@@ -37,9 +38,12 @@
     private String name;
     private String surname;
     private String phone;
-    private String address;
 
-    @ManyToMany(cascade = CascadeType.ALL)
+    @ManyToOne(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
+    @JoinColumn(name = "address_id")
+    private AddressEntity address;
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/model/db/CustomerEntity.java:44]: Ciekawi mnie sytuacja z tymi operacjami kaskadowymi. Co w przypadku gdy klient zostanie usunięty? Nie usuwamy adresów bo jeden adres może być do kilku osób?

File: backend/src/main/java/com/crm/service/impl/CarServiceImpl.java [hunk 3]
```diff
@@ -78,8 +78,6 @@ private boolean hasVinIllegalCharacters(final String vin) {
 
     @Override
     public CarResponse addCar(CarRequest carRequest) {
-
-        CarTypeEntity carTypeEntity;
         String nameOfCarTypeEntity = carRequest.getCarTypeEntity().getName();
 
         if (carRepository.findByRegistrationNumberIgnoreCase(carRequest.getRegistrationNumber()).isPresent()) {
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/service/impl/CarServiceImpl.java:83]: This two if's can be moved to one private method I think.

checkIfCarAlreadyExist()

just a suggestion

File: backend/src/main/java/com/crm/model/db/CustomerEntity.java (hunk 2 above)
Comment (aliwocha) [reply backend/src/main/java/com/crm/model/db/CustomerEntity.java:44]: Dokładnie tak to sobie obmyślłem. Wydaje mi się ok, ale jestem otwarty na sugestie.

File: backend/src/main/java/com/crm/model/db/AddressEntity.java (hunk 1 above)
Comment (aliwocha) [reply backend/src/main/java/com/crm/model/db/AddressEntity.java:47]: Haha, ok I will change that

File: backend/src/test-integration/java/com/crm/api/v1/CustomerControllerIT.java [hunk 4]
```diff
@@ -178,6 +195,20 @@ void shouldResponseWith_Created_WhenAddedCustomer() throws Exception {
                 .andExpect(status().is(CREATED.value()));
     }
 
+    @Transactional
+    @Rollback
```
Comment (SteelWro) [inline backend/src/test-integration/java/com/crm/api/v1/CustomerControllerIT.java:199]: I think @Transactional annotation on test has Rollback as default. @Rollback is not needed, but i'm not sure

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java [hunk 5]
```diff
@@ -61,13 +61,26 @@ public CustomerResponse addCustomer(final CustomerRequest customerRequest) {
         });
 
         CustomerEntity newCustomer = customerMapper.convertToEntity(customerRequest);
-        CustomerEntity savedCustomer = customerRepository.save(newCustomer);
+        return customerMapper.convertToDto(customerRepository.save(newCustomer));
+    }
+
+    private boolean isDuplicateCustomer(CustomerRequest customerRequest, CustomerEntity customerEntity) {
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:67]: final

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java [hunk 6]
```diff
@@ -61,13 +61,26 @@ public CustomerResponse addCustomer(final CustomerRequest customerRequest) {
...
+        return customerNameEquals(customerRequest, customerEntity)
+                && customerSurnameEquals(customerRequest, customerEntity)
+                && customerAddressEquals(customerRequest,customerEntity);
+    }
+
+    private boolean customerNameEquals(CustomerRequest customerRequest, CustomerEntity customerEntity) {
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:73]: final

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java [hunk 7]
```diff
@@ -61,13 +61,26 @@ public CustomerResponse addCustomer(final CustomerRequest customerRequest) {
...
+        return customerEntity.getName().equals(customerRequest.getName());
+    }
 
-        return customerMapper.convertToDto(savedCustomer);
+    private boolean customerSurnameEquals(CustomerRequest customerRequest, CustomerEntity customerEntity) {
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:77]: final

File: backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java [hunk 8]
```diff
@@ -61,13 +61,26 @@ public CustomerResponse addCustomer(final CustomerRequest customerRequest) {
...
+        return customerEntity.getSurname().equals(customerRequest.getSurname());
     }
 
-    private boolean isDuplicateCustomer(CustomerRequest customerRequest, CustomerEntity c) {
-        return c.getName().equals(customerRequest.getName()) && c.getSurname().equals(customerRequest.getSurname())
-                && c.getAddress().equals(customerRequest.getAddress());
+    private boolean customerAddressEquals(CustomerRequest customerRequest, CustomerEntity customerEntity) {
```
Comment (SteelWro) [inline backend/src/main/java/com/crm/service/impl/CustomerServiceImpl.java:81]: final

File: backend/src/test-integration/java/com/crm/api/v1/CustomerControllerIT.java (hunk 4 above)
Comment (aliwocha) [reply backend/src/test-integration/java/com/crm/api/v1/CustomerControllerIT.java:199]: I will check that

Comment (aliwocha) [reply backend/src/test-integration/java/com/crm/api/v1/CustomerControllerIT.java:199]: You might be right, thanks

File: backend/src/main/java/com/crm/service/impl/CarServiceImpl.java (hunk 3 above)
Comment (aliwocha) [reply backend/src/main/java/com/crm/service/impl/CarServiceImpl.java:83]: Good point, corrected

**Review by SteelWro — APPROVED** [review]

**Review by magikabdul — APPROVED** [review]

**Pull request merged**